docs/pam_otpasswd.8
.\"
.\"   pam_otpasswd(8) - One-Time Password Authentication System
.\"
.\"   Copyright (c) 2010-2013 Tomasz bla Fortuna
.\"
.\"   This file is part of OTPasswd.
.\"
.\"   OTPasswd is free software: you can redistribute it and/or modify
.\"   it under the terms of the GNU General Public License as published
.\"   by the Free Software Foundation, either version 3 of the License,
.\"   or any later version.
.\"
.\"   OTPasswd is distributed in the hope that it will be useful, but
.\"   WITHOUT ANY WARRANTY; without even the implied warranty of
.\"   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\"   GNU General Public License for more details.
.\"
.\"   You should have received a copy of the GNU General Public License
.\"   along with OTPasswd.  If not, see <http://www.gnu.org/licenses/>.
.\"
.\"   Author:  R Hannes Beinert & Tomasz bla Fortuna
.\"   Update:  02-Nov-13
.\"
.TH PAM_OTPASSWD 8 2013-11-02 "OTPasswd v0.8" "OTPasswd User Manual"
.\"
.SH NAME
pam_otpasswd \- Module for OTPasswd one-time password authentication
.\"
.SH SYNOPSIS
pam_otpasswd.so [\fIoptions\fR]
.\"
.\"  CONFIGURATION      [Normally only in Section 4]
.\"
.SH DESCRIPTION
This module allows the authentication of one-time passwords
with the \fBOTPasswd\fR system.
\fBOTPasswd\fR consists of two primary components, this PAM module
and the \fBotpasswd\fR(1) utility.
The utility is used to control the operation of the system,
and the PAM module essentially makes it happen.
.PP
In general, once \fBOTPasswd\fR has been installed,
the system administrator must determine where the system will keep
the user configuration data (user state).
The administrator will likely also configure the system by making various
policy choices,
such as how long passcodes must be, how complicated the static password should be,
which passcode alphabet will be used, etc.
Subsequently, users will probably also make some configuration choices
\- to the degree they are permitted by policy \-
and then they will generate a key and print some passcards.
All of these activities involve the use of the \fBotpasswd\fR(1) utility.
Once this initial configuration is complete,
the \fBOTPasswd\fR system will operate virtually autonomously.
Each time a user attempts to login to the system,
or otherwise authenticate with some service which uses this module,
the previously configured data is consulted,
a one-time passcode challenge will be issued,
and the user will have a chance to authenticate.
.PP
The \fBPAM\fR(7) authentication system supports four independent management
groups, of which this module only supports two:
\fBauth\fRentication management and \fBsession\fR management.
The \fBaccount\fR and \fBpassword\fR management groups are not supported.
.PP
In typical usage,
the \fBauth\fRentication component of the \fBpam_otpasswd\fR module
performs the task of issuing a challenge to the user for a one-time passcode,
validating it against the user state information, and advancing the
counter to the next passcode for subsequent authentications.
Other more advanced tasks involve the validation of the user's
\fIstatic password\fR and accepting some simple authentication-related
commands, such as sending a passcode to the user by text message (SMS).
.PP
The \fBsession\fR component of this module is primarily intended to
provide the user with warnings when a user session is initially established.
This include some of \fBOTPasswd\fR system conditions user should be aware of
like the fact that he is running out of printed passcodes
(see \fBotpasswd\fR(1) for more details).

This module's session management functions will never cause a
user session to be denied.
.PP
The \fBpam_otpasswd\fR module for compatibility with interfaces of
other modules may optionally be provided with a series of options
in the PAM configuration file (see EXAMPLE section below),
each of which may alter the behavior of some components of the
module.

This options duplicates functionality of otpasswd.conf, and enabling
any of them overrides settings in configuration file.

Any legal options which have no relevance to the specific management
context in which this module is called will be silently ignored.
Unrecognized options will be logged as warnings through \fBsyslog\fR(3).

.\"
.SH OPTIONS
.TP +4m
\fBdebug\fR
Maximize module \fBsyslog\fR(3) logging output.
While this is useful for developers, it is not typically what you want.
.\"
.TP +4m
\fBaudit\fR
Increased amount of logging to the system log.
This option is appropriate for system administrators who
desire to more carefully monitor \fBOTPasswd\fR activity
without excessive log spam.
.\"
.TP +4m
\fBsilent\fR
No unnecessary text or messages are sent to the calling program,
and the PAM conversation function is only used to ask for passcode.
Passing this option to \fBsession\fR stack effectively turns it off.
This option causes the module to behave in the same manner as though the
\fBPAM_SILENT\fR flag had been passed by the application.
.\"
.TP +4m
\fBuse_first_pass\fR
This option is ignored.
.\"
.PP
Invalid arguments are logged with \fBsyslog\fR(3),
but are otherwise ignored.
.\"
.SH MODULE SERVICES PROVIDED
\fBPAM\fR(7) supports four independent management groups, which are implemented as follows:
.PP
.PD 0
.RS
.TP +16m
\fBAuthentication\fR
Supported, except \fBpam_setcred\fR(3)
.TP +16m
\fBSession\fR
Fully supported
.TP +16m
\fBAccount\fR
Not supported
.TP +16m
\fBPassword\fR
Not supported
.RE
.PD
.\"
.\" *SECURITY NOTES
.\"  EXIT STATUS        [Normally only in Sections 1, 8]
.\"
.SH RETURN VALUES
.TP +4m
\fBPAM_SUCCESS\fR
User one-time passcode authentication was successful.
\fBpam_open_session\fR(3) will always return this value.
.TP +4m
\fBPAM_IGNORE\fR
Returned by \fBpam_sm_setcred\fR(3) when any attempt is made to alter
the user's credentials.
.\"
.TP +4m
\fBPAM_AUTH_ERR\fR
One of the following conditions has occurred:
.PD 0
.RS +4m
.IP \(bu +2m
User failed to properly provide the expected passcode
.IP \(bu +2m
\fBOTPasswd\fR use is enforced by policy but user has no configuration
.IP \(bu +2m
User state information was not successfully locked
.IP \(bu +2m
Passcode counter value could not be incremented
.IP \(bu +2m
Passcode prompt could not be generated
.IP \(bu +2m
Application provided no user response to a passcode query
.IP \(bu +2m
An unforeseen error occurred
.RE
.PD
.\"
.TP +4m
\fBPAM_USER_UNKNOWN\fR
The following conditions will result in this error:
.PD 0
.RS +4m
.IP \(bu +2m
System configuration file \fBotpasswd.conf\fR(5) is unreadable
.IP \(bu +2m
The \fBPAM\fR(7) system is unable to provide \fBpam_otpasswd\fR with
the name of the user for whom authentication has been requested.
See \fBpam_get_user\fR(3) for details.
.RE
.PD
.\"
.TP +4m
\fBPAM_SERVICE_ERR\fR
\fBOTPasswd\fR system configuration error.
This value is returned when any function in this PAM module is unable
to read the \fBotpasswd.conf\fR(5) system configuration file.
(FIXME: How does this return code translate into the success/failure of the auth stack?)
.\"
.PP
In those situations where an error status was returned to the calling
application, the \fBpam_otpasswd\fR module uses the \fBsyslog\fR(3)
facility to log additional details.
Check your system log.
.\"
.\"  ERRORS             [Typically only in Sections 2, 3]
.\"  ENVIRONMENT
.\"
.SH FILES
.TP
/etc/otpasswd
The directory for \fBOTPasswd\fR configuration files.
If \fBOTPasswd\fR is configured to use a global flat-file
database, then this file resides here, and therefore,
the directory should be owned by the special system user
(set in otpasswd.conf, a default value is \fIotpasswd\fR),
and should not be accessible for normal users
(mode 0700 recommended). The otpasswd(1) utility has to be
Set-UID to that user to be able to read and modify configuration and
state data.

Name of this directory can only be configured during compilation.
.\"
.TP
/etc/otpasswd/otpasswd.conf
The main \fBOTPasswd\fR system configuration file.
This file contains key-value configuration parameters that
govern the operation of the system.
In particular, it contains the state database configuration,
system-wide defaults, PAM configuration and policy configuration.
When the backend user database is configured to use MySQL or LDAP,
this file will contain privileged information and
must be not be accessible for normal users. This requires utility
to be Set-UID to user who owns /etc/otpasswd/otpasswd.conf.

Name of this file can only be configured during compilation.
.\"
.TP
/etc/otpasswd/otshadow
The system-wide user database, used only when the \fBotpasswd.conf\fR(5)
configuration file specifies the use of a global database backend.
The file contains state information for all users,
including keys, flags, etc, and must not be accessible
to normal users.
.\"
.TP
$HOME/.otpasswd
This file is only used when the system configuration file
\fBotpasswd.conf\fR(5) specifies that state information is
to be maintained in user home directories.
This has the same format as the \fBotshadow\fR(5) file above,
except it only contains information for a single user.

As users have full permissions to modify those files, policy can't
be enforced in this mode.
.\"
.TP
/etc/pam.d/otpasswd_login
The PAM prototype configuration for \fBOTPasswd\fR.
If this file is \fIinclude\fRd in a PAM configuration for any
service (like \fBSSH\fR), the \fBpam_otpasswd\fR(8) PAM module will be
used to ask the user for a passcode just after the normal \fBpam_unix\fR(8)
authentication mechanisms are applied. This stack is prepared to be used
instead of previous authentication state.
.\"
.\"  VERSIONS           [Normally only in Sections 2, 3]
.\" *COMPATIBILITY
.\"  CONFORMING TO
.\"
.SH NOTES
See \fBotpasswd\fR(1) for further information regarding the
\fBOTPasswd\fR one-time password authentication system.
In particular, the reader is directed to the sections entitled
COMPATIBILITY, DOCUMENTATION, and HISTORY.
.\"
.\"  BUGS
.\"
.SH EXAMPLE
A typical PAM \fBauth\fR stack for \fBOTPasswd\fR use would be:
.PP
.RS
.nf
auth      required    pam_tally.so onerr=succeed
auth      required    pam_shells.so
auth      required    pam_nologin.so
auth      required    pam_env.so
auth      requisite   pam_unix.so try_first_pass likeauth nullok

auth      required    pam_otpasswd.so audit
session   optional    pam_otpasswd.so
.fi
.RE
.PP
In this example, the \fIrequisite\fR keyword is used for the \fBpam_unix\fR(8)
module, which means that if the user fails to enter the proper system password,
the entire \fBauth\fR stack will fail immediately.
In particular, this means that the user will never be asked for a
passcode by the \fBpam_otpasswd\fR module.
To change this behavior, merely replace the \fIrequisite\fR keyword,
with the \fIrequired\fR keyword.
This forces the entire \fBauth\fR stack to complete first,
and the user will be asked for a passcode every time,
regardless of whether the user specified the system password correctly.
.PP
Note that the \fIaudit\fR option was specified in the \fBauth\fR configuration
for the \fBpam_otpasswd\fR module above.
This will cause an increased number of messages to be placed in the system log,
allowing system administrators to more
carefully monitor \fBOTPasswd\fR authentication activity.
.PP
This example also highlights the typical way in which \fBpam_otpasswd\fR
is used in the \fBsession\fR management stack.
The primary facility provided by this module's session management
functions is to provide warnings of unusual \fBOTPasswd\fR system
conditions when a user session is initially established
(see \fBotpasswd\fR(1) for more details).
(FIXME: verify previous)
No \fBpam_otpasswd\fR session management function will ever fail.
.\"
.SH SEE ALSO
\fBpam\fR(7),
\fBpam.conf\fR(5),
\fBotpasswd\fR(7),
\fBotpasswd\fR(1),
\fBotpasswd.conf\fR(5),
\fBotshadow\fR(5),
\fBotpasswd\fR(5)
.\"
.\" *DOCUMENTATION
.\" *AUTHORS
.\" *HISTORY
.\"
.SH LICENSE
Copyright (c) 2009, 2010 Tomasz bla Fortuna
.PP
This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.
.PP
This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.
.PP
You should have received a copy of the GNU General Public License
along with this program in a LICENSE file.
.\"
.SH AVAILABILITY
The latest version of the \fBOTPasswd\fR package is available in
source form at the project website
.nh
https://savannah.nongnu.org/projects/otpasswd
.hy 1
.\"
.\" End of Manual: pam_otpasswd(8)


ChangeLog
ChangeLog / TODO
================

[+] New functionality
[!] Fixed error
[%] Found bug
[*] Changes in existing functionality
[-] Feature in TODO
[?] New feature in TODO
[$] Idea which was given up

Trying to sort tasks according to their priority.

2013-10-23 0.8
	Preparing distributable, convoyable, usable and likable version.
	Big version jump because of quite a long time from previous changes.
	0.8 can be thought as a release candidate for 0.9 - I'm trying to keep
	versioning simple to ease packaging.

	* [!] In not mandatory mode will fail if user HOME is not found.
	* [%] PAM module maybe should mmap memory not to let key be additionaly swapped to disc.
	   It's is on the disc in plaintext but the less copies the better.

	* [*] By default running in DB=user. Trying to get it work out-of-the-box.
	* [*] Fixing default options to let otpasswd be more easily tested.
	* [*] Updated package name convention, new package script
	* [*] Move sources to src/ subdirectory.

	* [-] Approach to agent data passing is overcomplicated. stdin/stdout substitution
	      causes problems with communication and as it turns out - valgrind. Would better
	      keep stdin/stdout, but maybe not use them(?) Best to review.
        * [+] Review documentation and ensure it's up to date.
	* [+] Update Polish translation
	* [-] Relicense so it can be used more universally - will require changing of sha256
              implementation, but it would be best to use system openssl lib probably.
	* [+] Check how SUID version works.
	      Update: Was fine.
	* [-] Update defaults for OOB. Explain better how to configure OOB messaging.
	      Add otpasswd_oob checks to --check-config


2010-03-31 0.7_rc2
Done:
        * [!] On DB=user when administrator creates a key with -u option user
              can't access that state.
        * [!] Fixed security bug in static password handling.
        * [+] Dropping dependency on GMP. Cleaning up "num" interface.
          - Clean memory so there're no key/counter leaks
        * [+] Do not store things in mpz_t which don't need it (spass)
        * [+] Sanitize gettext environment
	* [+] Accept 2G[2] passcode specification entries.
	* [+] All previous functionality is now finally implemented.
	      All implemented testcases where successfully run after
	      The Split.
	* [+] Backward skipping removed from configuration.

        * [+] Printing with -l when skipped to the last passcard fixed .
        * [+] OOB Channel usage time updates + DoS security
        * [+] Ensure config privilege checking
        * [+] --info should print information whether static password is set.
        * [+] Write testcases for static password
        * [+] Make otpasswd work with su. Tested under Gentoo and works BUT
              without OOB (permission denied)
        * [+] "Problem solving section" in docs, mentioning use of -v on
              errors in the first place. This was done on the webpage.
        * [+] policy checking regarding salt. (won't fail sometimes)
              (Seems ok to me. Tested)

        * [$] Global DB should work with users which aren't in /etc/passwd.
              Should it?
        * [$] Add error_t and bool_t to differentiate differently
              used ints? (bool_t is very rarely used. It will better be left
              to documentation)
In progress:
        * [?] Cleanup PPP interface. This should look like follows.
              "State" is class implementing some basic features of
              state management. From this 'class' ppp is derived.
              ppp implements high-level functions which should be used
              explicitly to manage state information unless something
              more fine-grained is necessary. State on the other hand
              uses 'db' backend for operating on files/databases.
              * Lots of things done in direction to finish this task
        * [?] Security: Do we need to change something regarding ulimits?

TODO Major:
        * [?] Double passcode logging scheme
        * [%] Parameter -c alphabet=3,codelenght=8 is sometimes accepted
              but doesn't work.
        * [?] Parametrize PAM messages.
        * [?] Static password expire warnings + enforcement.
        * [?] Logging warnings printed to user at WARN level
        * [?] Key quality checking (duplicates)
        * [?] Check if lock files are links. if so. fail. Or rather always unlink
              before overwriting.
        * [?] Logging into syslog from utility if agent is SUID.
        * [?] Admin function to scan all states for things invalidating
              policy. This could be implemented in agent similar to
              --check-config
        * [-] Scan all FIXME/TODO entries (Especially OOB related)
        * [?] SELinux compatibility.
        * [?] Import/export of ascii state lines...
        * [?] Check if User can lock OTPasswd when working with su (stdout
        issue). We must NOT keep locks when outputing something to user.
        user
        * [?] OTPasswd when used with su should probably elevate it's rights
        with setuid() and then return back to original rights.
        * [?] I don't remember how I've done it and therefore I need to review
              security aspects of DB=user mode when user can mangle state file
              which is then accessed by uid0 PAM process.

Todo Minor:
        * [?] Will we be pppv3 compatible on big-endian?
        * [?] Sourced key generation (from file, string...)
        * [?] Improve LaTeX output (some colors, borders?)
        * [?] Incorporate SSH key fingerprints on passcards?
              Use some file as passcard background, allow hook
              regenerating this passcard
        * [-] Keep lines below 75 columns? Or 80?



2010-01-24 v0.5rc1
Done:
        * [+] Check bit distribution for alphabets not divisible by 2
        * [+] Remove dont-skip option.
        * [+] GMP might leak information with reallocs of it's mpz_t

pam/pam_otpasswd.c
/**********************************************************************
 * otpasswd -- One-time password manager and PAM module.
 * Copyright (C) 2009, 2010 by Tomasz bla Fortuna <bla@thera.be>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with otpasswd. If not, see <http://www.gnu.org/licenses/>.
 **********************************************************************/

#define PAM_SM_AUTH
#define PAM_SM_SESSION
#define _PAM_EXTERN_FUNCTIONS

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h> /* sleep */

/* libotp interface */
#include "ppp.h"

/* PAM declarations */
#include <pam_modules.h>
#include "pam_helpers.h"

/** Entry point for authentication */
PAM_EXTERN int pam_sm_authenticate(
	pam_handle_t *pamh, int flags, int argc, const char **argv)
{
	int retval;

	/* Static password failure delay */
	const int spass_delay = 2;

	/* Prompt to ask user */
	const char *prompt = NULL;

	/* Required for communication with user */
	struct pam_response *resp = NULL;

	/* OTP State */
	state *s = NULL;

	/* Parameters */
	cfg_t *cfg = NULL;

	/* Username */
	const char *username = NULL;

	/* Have user entered OOB in this session? */
	int oob_sent = 0;
	
	/* User messages */
	const char *oob_msg = "Out-of-band message sent.";
	const char *oob_already_msg = "Out-of-band message already sent.";

	/* Counters required for login algorithm */
	int first_try = 1;
	int dont_increment = 0; /* Do not increment if previous prompt was for OOB */
	int tries;


	/* Perform initialization:
	 * parse options, start logging, initialize state,
	 */
	retval = ph_init(pamh, flags, argc, argv, &s, &username);
	if (retval != 0)
		return retval;

	cfg = cfg_get();

	if (cfg->pam_spass_require == CONFIG_ENABLED) {
		/* Before we will enter passcode loop ask user for his
		 * static password. As this is supposed to be used instead
		 * of unix password we behave similarly. We ask questions
		 * even if used doesn't have state and don't tell anything. */
		int loaded = 0;
		retval = ppp_state_load(s, PPP_DONT_LOCK);
		if (retval == 0)
			loaded = 1;
		else
			print(PRINT_NOTICE,
			      "unable to read state when "
			      "asking for static password; user=%s error=%d\n",
			      username, retval);

		if (ph_validate_spass(pamh, loaded ? s : NULL, username) != 0) {
			sleep(spass_delay);
			return PAM_AUTH_ERR;
		}
	}


	first_try = 1;
	dont_increment = 0;
	for (tries = 0; tries < (cfg->pam_retry == 0 ? 1 : cfg->pam_retries);) {
		if (first_try || cfg->pam_retry == 1) {
			/* First time or we are retrying while changing the passcode */
			first_try = 0;
			if (dont_increment) 
				dont_increment = 0;
			else {
				retval = ph_increment(pamh, username, s);
				if (retval != 0)
					goto cleanup;

				/* Generate fresh prompt */
				retval = ppp_get_str(s, PPP_FIELD_PROMPT, &prompt);
				if (retval != 0 || !prompt) {
					print(PRINT_ERROR, "error while generating prompt; user=%s", 
					      username);
					retval = PAM_AUTH_ERR;
					goto cleanup;
				}
			}
		}

		/* If user configurated OOB to be send
		 * all the time - sent it */
		if (cfg->pam_oob == OOB_ALWAYS) {
			ph_oob_send(pamh, s, username);
		}

		retval = PAM_AUTH_ERR;

		/* State question about passcode. User might
		 * request OOB by answering with '.'. If so
		 * we perform OOB and restate question */
		resp = ph_query_user(pamh,
				     ppp_flag_check(s, FLAG_SHOW),
				     prompt);

		if (!resp) {
			/* No response? */
			print(PRINT_NOTICE, "no response from user during auth; user=%s\n", 
			      username);
			goto cleanup;
		}

		/* Hook up OOB request */
		if (strcmp(resp->resp, ".") == 0) {
			/* Drop reply. We will most probably 
			 * restate question about passcode. */
			ph_drop_response(resp);

			/* Was it already sent? */
			if (oob_sent) {
				/* if so - ignore prompt with message */
				ph_show_message(pamh, oob_already_msg, username);
				dont_increment = 1;
				continue;
			}

			/* Only if not already sent in this session. */
			switch (cfg->pam_oob) {
			case OOB_REQUEST:
				if (ph_oob_send(pamh, s, username) == 0) {
					ph_show_message(pamh, oob_msg, username);
					oob_sent = 1;
				}
				break;
				
			case OOB_SECURE_REQUEST:
				if (ph_validate_spass(pamh, s, username) == 0) {
					if (ph_oob_send(pamh, s, username) == 0) {
						ph_show_message(pamh, oob_msg, username);
						oob_sent = 1;
					}
				} else {
					/* Ensure attacker doesn't have 
					 * infinite tries of static pass */
					sleep(spass_delay);
					tries++;
				}
				break;
			default:
				print(PRINT_ERROR, "Internal error: "
				      "Invalid option read from config file.\n");
				assert(0);
				retval = PAM_AUTH_ERR;
				goto cleanup;
			}

			/* Continue, so the user is restated question about passcode */
			dont_increment = 1;
			continue;
		}


		/* Count this try */
		tries++;

		if (ppp_authenticate(s, resp[0].resp) == 0) {
			/* Authenticated */
			ph_drop_response(resp);

			/* Correctly authenticated */
			retval = PAM_SUCCESS;

			print(PRINT_WARN,
			      "accepted otp authentication; user=%s\n", username);
			goto cleanup;
		}

		ph_drop_response(resp);

		/* Increment count of failures */
		retval = ppp_failures(s, 0);
		if (retval != 0) {
			print(PRINT_WARN, "unable to increment failure count; user=%s", 
			      username);
		}

		/* Error during authentication */
		retval = PAM_AUTH_ERR;

		print(PRINT_WARN, 
		      "authentication failure; user=%s; try=%d/%d\n",
		      username, tries+1, cfg->pam_retries);
	}

cleanup:
	ph_fini(s);
	return retval;
}

PAM_EXTERN int pam_sm_open_session(
	pam_handle_t *pamh, int flags, int argc, const char **argv)
{
	int retval;

	/* Should we store state after printing? */
	int release_flags = 0;

	/* OTP State */
	state *s = NULL;

	/* Username */
	const char *username = NULL;

	/* User warning conditions */
	int err;

	/* Initialize */
	retval = ph_init(pamh, flags, argc, argv, &s, &username);
	if (retval != 0) {
		print(PRINT_WARN, "error while initializing PPP; user=%s\n", username);
		return retval;
	}

	print(PRINT_NOTICE, "session entrance; user=%s\n", username);

	if (ppp_state_load(s, 0) != 0)
		goto exit;

	print(PRINT_NOTICE, "state loaded; user=%s\n", username);

	err = ppp_get_warning_conditions(s);
	if (err == 0) {
		/* No warnings! */
		print(PRINT_NOTICE, "no warning to print; user=%s\n", username);
		goto cleanup;
	}


	{
		const char *msg = NULL;
		int err_copy = err;
		char buff_msg[300] = {0};
		int len;

		while ((msg = ppp_get_warning_message(s, &err_copy)) != NULL) {
			/* Generate message */

			len = snprintf(buff_msg, sizeof(buff_msg), "*** OTP Warning: %s", msg);
			if (len < 10) {
				print(PRINT_ERROR, "internal error: strange sprintf error; user=%s\n", username);
				goto cleanup;
			}

			ph_show_message(pamh, buff_msg, username);
		}
	}

	/* Have we printed warning about recent failures? */
	if (err & PPP_WARN_RECENT_FAILURES) {
		if (ppp_set_int(s, PPP_FIELD_RECENT_FAILURES, 0, PPP_CHECK_POLICY) != 0)
			print(PRINT_WARN, "unable to clear recent failures; user=%s\n", username);
		release_flags = PPP_STORE;
	}

cleanup:
	ppp_state_release(s, PPP_UNLOCK | release_flags);
exit:
	ph_fini(s);

	/* Ignore us, even if we fail. */
	return PAM_IGNORE;
}


/* Ignored/not-implemented functions */
PAM_EXTERN int pam_sm_setcred(
	pam_handle_t *pamh, int flags, int argc, const char **argv)
{
	 return PAM_IGNORE;
}

PAM_EXTERN int pam_sm_close_session(
	pam_handle_t *pamh, int flags, int argc, const char **argv)
{
	return PAM_IGNORE;
}

#ifdef PAM_STATIC

/* Initialization struct */
struct pam_module _pam_permit_modstruct = {
	"otpasswd",
	otp_authenticate,
	otp_setcred,
	NULL,
	otp_open_session,
	otp_close_session,
	NULL
};

#endif
